Completed Google Form script generator

esfoobar-usds opened the pull request:
This Google Form Script can be dropped on any Google Drive and run to generate the Justice 40 ingestion Form.

Pending work: assign a Google Sheet by default.

lucasmbrown-usds commented:
This is the same issue we were talking about earlier today, but I get nervous about having to maintain duplicate copies of the same schema. As I see it now, we've got one copy of the schema in the `data_set_description_schema.yaml`, and then one copy is spread throughout the source code (e.g. where we write the options such as "Esri Shapefile"). 

Is it possible to autogenerate those options from the schema itself, so we don't have to update the options in two places everytime there's a change, we just run a script? 

Thanks!

esfoobar-usds commented:
> This is the same issue we were talking about earlier today, but I get nervous about having to maintain duplicate copies of the same schema. As I see it now, we've got one copy of the schema in the `data_set_description_schema.yaml`, and then one copy is spread throughout the source code (e.g. where we write the options such as "Esri Shapefile").

> Is it possible to autogenerate those options from the schema itself, so we don't have to update the options in two places everytime there's a change, we just run a script?

Yes. I mentioned on the meeting that we would do this initial phase first of having the script file on the repo so that we can have the form by Monday and then next week start to work on a Python script to generate the Google Script programatically from the YAML definition file.

lucasmbrown-usds commented:
> Yes. I mentioned on the meeting that we would do this initial phase first of having the script file on the repo so that we can have the form by Monday and then next week start to work on a Python script to generate the Google Script programatically from the YAML definition file.

Ok. I'm also a little nervous that this is only a create script; after it's run the very first time, what is our plan for how we'll run this again? Are we able to use create or update functionality so that you can update the same form "in place", or are we planning to distribute a new link each time the form is updated (perhaps obscured behind a link shortener or something)? 

Thanks! Overall looking good!

esfoobar-usds commented:
> Ok. I'm also a little nervous that this is only a create script; after it's run the very first time, what is our plan for how we'll run this again? Are we able to use create or update functionality so that you can update the same form "in place", or are we planning to distribute a new link each time the form is updated (perhaps obscured behind a link shortener or something)?

[Per the caveats here](https://app.zenhub.com/workspaces/justice40-60993f6e05473d0010ec44e3/issues/usds/justice40-tool/39#issuecomment-846159519), there is no way that I've found where the script run can update the same form or point to the same sheet, so these are factors that push me to believe we're going to run into issues pretty quickly, unless we commit to creating a proper submission app.

esfoobar-usds commented:
Hi all, I have updated the script with the following changes:

- The script now generates one single Google Form and updates it
- The script writes to one single Google Sheet, although it creates a new Sheet (tab) every time the script is run
- Added descriptions to all fields
- Added instructions on Google Sheet

You can q/a the new form [here](https://docs.google.com/forms/d/e/1FAIpQLSe9EvOe90OWyrBVGo5QbCQX9Orz3lOjWi9rgBAewUQUb6PFoA/viewform).

lucasmbrown-usds commented on data-roadmap/data_roadmap/data_set_description_schema.yaml:
```diff
@@ -2,23 +2,26 @@
-  required=True)
-sponsor: str(required=True)
+data_formats:
+  enum('GeoJSON', 'Esri Shapefile (SHP, DBF, SHX)', 'GML', 'KML/KMZ',
```
Nice cleanup!

switzersc-usds replied:
We need to add data formats to this, eg TXT. Maybe this shouldn't be an enum yet.

lucasmbrown-usds commented on data-roadmap/data_roadmap/data_set_descriptions/google_form_gen.js:
```diff
@@ -1,15 +1,31 @@
 function createForm() {
+  var form = FormApp.openByUrl(
+    "https://docs.google.com/forms/d/1Kb9Hl82I9kPJhqS_3bQDbWWvzCxPDTZBaI7_Eo6339E/edit"
```
Very nice!

ginabeena commented:
Re: schema -- we could easily have a [type provider](https://docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/) or [protobuf](https://developers.google.com/protocol-buffers) format and hook host the schema registry. Would be a high value add early on that will amortize work and provide high ROI. That way you don't have stringly type inconsistencies, you can version it, you can do some binary lensing into header values, etc.

switzersc-usds commented on data-roadmap/data_roadmap/data_set_description_field_descriptions.yaml:
```diff
@@ -1,39 +1,23 @@
 # There is no method for adding field descriptions to `yamale` schemas.
 # Therefore, we've created a dictionary here of fields and their descriptions.
-name: A short name of the data set.
```
We've ended up updating the descriptions directly in the Google spreadsheet. We should update this yaml file with those descriptions.

switzersc-usds reviewed (changes requested):
We should update the YAML field descriptions with the latest from the sheet.

switzersc-usds commented:
I will look into this a bit more in depth before sprint planning to see what is left here and how to update the card.

switzersc-usds commented:
Closing this as we have removed the data-roadmap directory for now. We can revisit this feature set at a later date.
